skimage/util/_flood_fill.py

```diff
@@ -0,0 +1,186 @@
+                                      _set_edge_values_inplace, _fast_pad)
+    from ..morphology.watershed import _offsets_to_raveled_neighbors
+    from ._flood_fill_cy import _flood_fill_equal, _flood_fill_tolerance
+
+    image = np.asarray(image, order='C')
+    seed_value = image[seed_point]
+
+    # Shortcut for rank zero
+    if image.size == 0:
+        return np.array([], dtype=(np.intp if indices else np.uint8))
+
+    # Convenience for 1d input
+    try:
+        iter(seed_point)
+    except TypeError:
+        seed_point = (seed_point,)
+
+    selem = _resolve_neighborhood(selem, connectivity, image.ndim)
+
+    # Must annotate borders
+    working_image = _fast_pad(image, image.min())
+
+    # Correct start point in ravelled image
+    ravelled_seed_idx = np.ravel_multi_index([i+1 for i in seed_point],
+                                             working_image.shape)
+    neighbor_offsets = _offsets_to_raveled_neighbors(
+        working_image.shape, selem, center=((1,) * image.ndim))
+
+    # Use a set of flags; see _flood_fill_cy.pyx for meanings
+    flags = np.zeros(working_image.shape, dtype=np.uint8)
+    _set_edge_values_inplace(flags, value=2)
+
+    try:
+        if tolerance is not None:
+            # Check if tolerance could create overflow problems
+            try:
+                max_value = np.finfo(working_image.dtype).max
+                min_value = np.finfo(working_image.dtype).min
+            except ValueError:
+                max_value = np.iinfo(working_image.dtype).max
+                min_value = np.iinfo(working_image.dtype).min
+
+            high_tol = min(max_value, seed_value + tolerance)
+            low_tol = max(min_value, seed_value - tolerance)
+
+            _flood_fill_tolerance(working_image.ravel(), 
+                                  flags.ravel(), 
+                                  neighbor_offsets,
+                                  ravelled_seed_idx, 
+                                  seed_value, 
+                                  low_tol, 
+                                  high_tol)
+        else:
+            _flood_fill_equal(working_image.ravel(), 
+                              flags.ravel(), 
+                              neighbor_offsets,
+                              ravelled_seed_idx, 
+                              seed_value)
+    except TypeError:
+        if working_image.dtype == np.float16:
+            # Provide the user with clearer error message
+            raise TypeError("dtype of `image` is float16 which is not "
+                            "supported, try upcasting to float32")
+        else:
+            raise
+
+    # Output what the user requested
+    original_slice = (slice(1, -1),) * image.ndim
+    if indices:
+        return np.nonzero(flags[original_slice] == 1)
+    else:
+        if not inplace:
+            output = working_image[original_slice]
+            output[flags[original_slice] == 1] = new_value
```

Comment:
Just an idea: Have you experimented with setting `new_value` inside Cython? I not sure whether it would be faster but would be certainly interested in the result. :)

Reply:
That's a good point, and I haven't.  If it works and has some benefit, might be worth considering refactoring to `_shared` because this is a really common operation.

Reply:
Gave it a shot, there was not a significant difference.